### PR TITLE
Brodes/seh flow phase2 splitting seh edges

### PR DESCRIPTION
**Pull request by bdrodes**
Phase 2 SEH handling PR. This phase simply establishes there are two edge types. Where a specific edge must be chosen, we assume it is a C++ edge. The next phase PR will fully establish uses of these edge types to correctly distinguish SEH from ordinary C++ exception edges.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 1]
```diff
@@ -3,13 +3,13 @@
...
 private import internal.EdgeKindInternal
-private import codeql.util.Boolean
 
 private newtype TEdgeKind =
   TGotoEdge() or // Single successor (including fall-through)
   TTrueEdge() or // 'true' edge of conditional branch
   TFalseEdge() or // 'false' edge of conditional branch
-  TExceptionEdge(Boolean isSeh) or // Thrown exception, true for SEH exceptions, false otherwise
+  TCppExceptionEdge() or // Thrown C++ exception
+  TSehExceptionEdge() or // Thrown C++ exception
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:12]:
```suggestion
  TSehExceptionEdge() or // Thrown SEH exception
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 2]
```diff
@@ -52,23 +52,33 @@ class FalseEdge extends EdgeKindImpl, TFalseEdge {
...
-  ExceptionEdge() { this = TExceptionEdge(isSeh) }
+final class ExceptionEdge = ExceptionEdgeImpl;
 
-  /**
-   * Holds if the exception is a Structured Exception Handling (SEH) exception.
-   */
-  final predicate isSeh() { isSeh = true }
+/**
+ * An "exception" edge, representing the successor of an instruction when that
+ * instruction's evaluation throws an exception for C++ exceptions
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:70]:
```suggestion
 * instruction's evaluation throws a C++ exception.
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 3]
```diff
@@ -52,23 +52,33 @@ class FalseEdge extends EdgeKindImpl, TFalseEdge {
...
+class CppExceptionEdge extends ExceptionEdgeImpl, TCppExceptionEdge {
+  final override string toString() { result = "C++ Exception" }
+}
 
-  final override string toString() {
-    if isSeh = true then result = "SEH Exception" else result = "C++ Exception"
-  }
+/**
+ * An "exception" edge, representing the successor of an instruction when that
+ * instruction's evaluation throws an exception for SEH exceptions
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:78]:
```suggestion
 * instruction's evaluation throws an SEH exception.
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 4]
```diff
@@ -123,9 +145,22 @@ module EdgeKind {
   FalseEdge falseEdge() { result = TFalseEdge() }
 
   /**
-   * Gets the single instance of the `ExceptionEdge` class.
+   * Gets an instance of the `CppExceptionEdge` class.
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:148]:
```suggestion
   * Gets the single instance of the `CppExceptionEdge` class.
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 5]
```diff
@@ -123,9 +145,22 @@ module EdgeKind {
...
+   */
+  CppExceptionEdge cppExceptionEdge() { result = TCppExceptionEdge() }
+
+  /**
+   * Gets an instance of the `SehExceptionEdge` class.
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:153]:
```suggestion
   * Gets the single instance of the `SehExceptionEdge` class.
```

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 6]
```diff
@@ -123,9 +145,22 @@ module EdgeKind {
...
-  ExceptionEdge exceptionEdge() { result = TExceptionEdge() }
+  SehExceptionEdge sehExceptionEdge() { result = TSehExceptionEdge() }
+
+  /**
+   * Gets an instance of the `ExceptionEdge` class.
+   */
+  ExceptionEdge exceptionEdge() {
+    result = cppExceptionEdge() or
+    result = sehExceptionEdge()
+  }
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:163]: I would not introduce this predicate, unless you need it somewhere later. I don't think it's being used right now, correct?

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll [hunk 7]
```diff
@@ -51,12 +52,33 @@ class FalseEdge extends EdgeKindImpl, TFalseEdge {
...
+abstract private class ExceptionEdgeImpl extends EdgeKindImpl { }
+
 /**
  * An "exception" edge, representing the successor of an instruction when that
  * instruction's evaluation throws an exception.
+ *
+ * Exception edges are expclitly sublcassed to
+ * `CppExceptionEdge` and `SehExceptionEdge` only.
+ * Further sublcasses, if required, should be added privately
+ * here for IR efficiency.
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:64]: Nit picking to give the lines similar lengths as the earlier ones:
```suggestion
 * Exception edges are expclitly sublcassed to `CppExceptionEdge` and `SehExceptionEdge`
 * only. Further sublcasses, if required, should be added privately here for IR efficiency.
```

**Review by jketema — COMMENTED** [review]
Mostly comments about the comments again. Just one question about the newly introduced predicate.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll (hunk 1 above)
Comment (bdrodes) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:12]: COPY PASTA!!... grr sorry about that.

File: cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll (hunk 6 above)
Comment (bdrodes) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll:163]: Only introduced to keep parity with what we had. I'll delete it.

Comment (bdrodes) [reply cpp/ql/lib/semmle/code/cpp/ir/implementation/EdgeKind.qll]: Done

**Review by jketema — COMMENTED** [review]
LGTM. Let me run some DCA on this, once that's done and looks good, I'll merge.

**Pull request merged**